sprint_11

alekseiypetrov opened the pull request.

Commit:
Добавлен индикатор загрузки при авторизации.

Commit:
Добавлена логика блокировки UI при авторизации.
Реализована синхронизация SplashScreen через NotificationCenter из-за возникающей гонки, когда SplashScreenViewController.viewDidAppear(animated) выполнялся раньше, чем успевал выполниться запрос и, соответственно, сохранялся токен, что вызывало зацикливание: SplashScreen -> AuthScreen -> WebScreen -> Авторизация -> SplashScreen и Выполнение POST-запроса (асинхронно) -> AuthScreen (независимо от наличия/отсутствия токена в хранилище).

Commit:
1) В Constants перенесены настройка декодера JSON и расширение для отслеживания о получении и сохранения токена.
2) Класс OAuth2TokenStorage был переделан под паттерн Singleton. Аргументация: SplashViewController, ProfileViewController и OAuth2Service используют данный класс для получения токена (первые 2 класса) и для сохранения токена (последний класс). Я посчитал целесообразным, что бы объект для хранилища был един.
3) Добавлен класс ProfileService, который обеспечивает загрузку базовой информации о профиле из сети. Соответственно, еще был переписан метод настройки UI в ProfileViewController.

Commit:
Исправлена логика создания и обновления полей лейблов на экране профиля.

Commit:
Убрана логика с NotificationCenter для отслеживания токена.
Изменена логика при отправке POST-запроса при авторизации.

Commit:
Изменена логика поведения SplashScreen в случае, если токен уже есть.
Реализовано обновление UI ProfileView после загрузки данных профиля пользователя.

Commit:
Реализован ProfileImageService, который загружает URL аватарки пользователя.
Добавлено отслеживание уведомления об успешной загрузке URL аватарки в NotificationCenter.

Commit:
Выполнена замена API KVO для WKWebView.estimatedProgress.
Имплементировано в расширении URLSession дженерик-метод objectTask(for:) для уменьшения дублирования кода, связанного с обработкой результатов запроса в data(for:).
Заменено использование data(for:) на objectTask(for:).
При получении ошибки в сетевом запросе в AuthViewController реализовано отображение UIAlertController с заголовком «Что-то пошло не так», подписью «Не удалось войти в систему» и кнопкой «Ок» для скрытия алерта.

Commit:
Добавлен фреймворк Kingfisher и реализована загрузка изображения профиля.
Возможно, стоит рассмотреть загрузку не small изображения, а large, чтобы отображени картинки выглядела лучше.

Commit:
Добавлен SwiftKeychainWrapper в проект через SPM.
В OAuth2TokenStorage изменено хранение и получение из хранилища токена с использованием SwiftKeychainWrapper вместо UserDefaults.

Commit:
Выполнена создание кодом ProfileViewController'а.

Commit:
Выполнена верстка SplashViewController.
Важная заметка: по заданию в практикуме предлагалось из SplashViewController взаимодействовать напрямую AuthViewController и переходить к нему, что, вообще говоря, нарушило бы немного дизайн по нашему макету. Я немного видоизменил код, я оставил создание AuthViewController, назначил ему делегата, но именно функцию present() я применяю к navigationController.

d00m1r commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -32,25 +33,40 @@ final class AuthViewController: UIViewController {
                 print("Авторизация прошла успешно, начинается перенаправление к галерее.")
-                self.delegate?.didAuthentificate(self)
+                vc.dismiss(animated: true) { [weak self] in
+                    guard let self = self else { return }
```
Более лаконичный вариант
`guard let myVar else { return }`
`if let myVar { ... }`
Того же касается и `self`
Далее по аналогии

d00m1r commented on ImageFeed/Constants.swift:
```diff
@@ -7,3 +7,10 @@ enum Constants {
     static let defaultBaseURL = URL(string: "https://api.unsplash.com")
 }
+
+class SnakeCaseJSONDecoder: JSONDecoder {
```
Следует добавлять модификатор final к классам, которые не планируются быть наследуемыми
Во-первых, это упрощает код, визуально понятно, что класс не является наследуемым
Во-вторых, это оптимизирует работу приложения, о чем ты сможешь узнать в дальнейших уроках

d00m1r commented on ImageFeed/ServiceAccessLayer/GetProfile/ProfileImageService.swift:
```diff
@@ -0,0 +1,70 @@
+import Foundation
+
+struct ProfileImage: Codable {
```
1 файл = 1 сущность
Лучше разделять сущности на файлы, для удобства навигации и изменения. Модель в данном случае лучше вынести в слой моделей, исключением может являться создание приватных типов внутри других

d00m1r commented on ImageFeed/SplashViewController/SplashViewController.swift:
```diff
@@ -1,47 +1,85 @@
+                return
+            }
+            authViewController.delegate = self
+//            authViewController.modalPresentationStyle = .fullScreen
```
Давай поддерживать код в чистоте - уберем весь закомментированный код

d00m1r commented on ImageFeed/ServiceAccessLayer/Network/URLSession+data.swift:
```diff
@@ -19,14 +19,14 @@ extension URLSession {
             }
             if let response = response,
                let statusCode = (response as? HTTPURLResponse)?.statusCode,
                statusCode < 200 || statusCode >= 300 {
```
Проверять диапозон удобнее используя [оператор `~=`](https://stackoverflow.com/questions/38371870/operator-in-swift)

Commit:
Внесены изменения по комментариям в github.

Commit:
Внесены изменения по комментариям из практикума.
Убрано использования force unwrap в ImagesListCell и ProfileViewController.
Решена ошибка декодирования (только сегодня узнал, что при регистрации обязательным параметром является только username, а остальные опциональные), поэтому не полностью заполненный профиль также может быть загружен. Также была изменена фотография, которая загружается на экране профиля (large выглядит гораздо четче, чем small, до этого были одни пиксели).

Commit:
Исправлена настройка цвета TabBar (теперь как по макету).
Приняты во внимание рекомендации по верстке кодом в ProfileViewController (все вью вынесены в private lazy var, декомпозиция методов и вынесение постоянных значений в enum Constants).
